From b0f32fd47d2d5043834a9d64a70d1801df04fc16 Mon Sep 17 00:00:00 2001
From: Guan Xuetao <gxt@pku.edu.cn>
Date: Thu, 17 Sep 2026 11:29:22 +0000
Subject: [PATCH 1/6] DADAO: Register dadao triple and add to LLVM_ALL_TARGETS

Register the DADAO target triple in LLVM's Triple infrastructure:
- Triple.h: Add dadao to ArchType enum, update LastArchType
- Triple.cpp: Add dadao to getArchTypeName, getArchTypePrefix,
  getArchTypeForLLVMName, parseArch, getDefaultFormat (ELF),
  getArchPointerBitWidth (64-bit)
- DADAO is big endian (not in isLittleEndian), so defaults to false
- CMakeLists.txt: Add DADAO to LLVM_ALL_TARGETS (ADR-0007)

Triple: dadao-unknown-elf
Pointer width: 64-bit
Endianness: big endian (ELFDATA2MSB)
Default object format: ELF
---
 llvm/CMakeLists.txt                     |  1 +
 llvm/include/llvm/TargetParser/Triple.h |  3 ++-
 llvm/lib/TargetParser/Triple.cpp        | 10 ++++++++++
 3 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/llvm/CMakeLists.txt b/llvm/CMakeLists.txt
index 9340cd6c4..7789dd768 100644
--- a/llvm/CMakeLists.txt
+++ b/llvm/CMakeLists.txt
@@ -594,6 +594,7 @@ set(LLVM_ALL_TARGETS
   ARM
   AVR
   BPF
+  DADAO
   Hexagon
   Lanai
   LoongArch
diff --git a/llvm/include/llvm/TargetParser/Triple.h b/llvm/include/llvm/TargetParser/Triple.h
index d819c4975..ad3866972 100644
--- a/llvm/include/llvm/TargetParser/Triple.h
+++ b/llvm/include/llvm/TargetParser/Triple.h
@@ -114,7 +114,8 @@ public:
     renderscript32, // 32-bit RenderScript
     renderscript64, // 64-bit RenderScript
     ve,             // NEC SX-Aurora Vector Engine
-    LastArchType = ve
+    dadao,          // DADAO: SimRISC (big endian)
+    LastArchType = dadao
   };
   enum SubArchType {
     NoSubArch,
diff --git a/llvm/lib/TargetParser/Triple.cpp b/llvm/lib/TargetParser/Triple.cpp
index 356782c1f..838c24fac 100644
--- a/llvm/lib/TargetParser/Triple.cpp
+++ b/llvm/lib/TargetParser/Triple.cpp
@@ -152,6 +152,8 @@ StringRef Triple::getArchTypeName(ArchType Kind) {
     return "thumbeb";
   case ve:
     return "ve";
+  case dadao:
+    return "dadao";
   case wasm32:
     return "wasm32";
   case wasm64:
@@ -415,6 +417,9 @@ StringRef Triple::getArchTypePrefix(ArchType Kind) {
 
   case xtensa:
     return "xtensa";
+
+  case dadao:
+    return "dadao";
   }
 }
 
@@ -567,6 +572,7 @@ Triple::ArchType Triple::getArchTypeForLLVMName(StringRef Name) {
       .Case("loongarch64", loongarch64)
       .Case("dxil", dxil)
       .Case("xtensa", xtensa)
+      .Case("dadao", dadao)
       .Default(UnknownArch);
 }
 
@@ -721,6 +727,7 @@ Triple::ArchType Triple::parseArch(StringRef ArchName) {
                   "dxilv1.9"},
                  Triple::dxil)
           .Case("xtensa", Triple::xtensa)
+          .Case("dadao", Triple::dadao)
           .Default(Triple::UnknownArch);
 
   // Some architectures require special parsing logic just to compute the
@@ -1046,6 +1053,7 @@ static Triple::ObjectFormatType getDefaultFormat(const Triple &T) {
   case Triple::ve:
   case Triple::xcore:
   case Triple::xtensa:
+  case Triple::dadao:
     return Triple::ELF;
 
   case Triple::mipsel:
@@ -1839,6 +1847,7 @@ unsigned Triple::getArchPointerBitWidth(llvm::Triple::ArchType Arch) {
   case llvm::Triple::ve:
   case llvm::Triple::wasm64:
   case llvm::Triple::x86_64:
+  case llvm::Triple::dadao:
     return 64;
   }
   llvm_unreachable("Invalid architecture value");
@@ -2198,6 +2207,7 @@ Triple Triple::getLittleEndianArchVariant() const {
 
   switch (getArch()) {
   case Triple::UnknownArch:
+  case Triple::dadao:
   case Triple::lanai:
   case Triple::sparcv9:
   case Triple::systemz:
